### PR TITLE
Add project delete command

### LINKED ISSUE #646: Is there a way to delete project with odo?
`odo project' has `create`, but no `delete` command. Is that done for purpose?

### PR DESCRIPTION
**Pull request by anmolbabu**
This PR adds project delete command.
The core implementation is only wrapped around by this PR
and exposed to the outside as a cli command

fixes #646
Signed-off-by: anmolbabu <anmolbudugutta@gmail.com>

### CODE REVIEW

File: pkg/project/project.go [hunk 1]
```diff
@@ -34,6 +34,14 @@ func Create(client *occlient.Client, projectName string) error {
 	return nil
 }
 
+func Delete(client *occlient.Client, projectName string) error {
```
Comment (syamgk) [inline pkg/project/project.go]: It will be good to add unit test for this

Comment (surajnarwade) [reply pkg/project/project.go]: +1

File: cmd/project.go [hunk 2]
```diff
@@ -105,6 +106,23 @@ var projectCreateCmd = &cobra.Command{
 	},
 }
 
+var projectDeleteCmd = &cobra.Command{
+	Use:   "delete",
+	Short: "Delete a project",
+	Long:  "Delete a project",
```
Comment (cmoulliard) [inline cmd/project.go]: What do you plan to delete ? 
- All the k8's/openshift resources added to a namespace or not and selected using k8s label
- The current namespace
What is a project ? Can you improve the description of the command and what will be done please ?

**Comment by anmolbabu** [comment]
ToDo(@anmolbabu)

1. Handle the currentproject in case when the only project is being deleted..
2. Add UTs and e2e tests
3. Address comments on the PR

**Comment by surajnarwade** [comment]
@anmolbabu , go through `oc delete project <project_name>`, see how `oc` handles corner cases

File: pkg/project/project.go (hunk 1 above)
Comment (anmolbabu) [reply pkg/project/project.go]: > It will be good to add unit test for this

Ack Done

> unit test for this, https://github.com/redhat-developer/odo/blob/9cb001ec7d86d39d06084c0247a9cf762824323e/pkg/occlient/occlient.go#L1151

Func remains Unaltered by me :)

File: cmd/project.go (hunk 2 above)
Comment (anmolbabu) [reply cmd/project.go]: @cmoulliard I am changing the long description to "Delete a project and all resources deployed in the project being deleted" 

I am not sure namespace as a term is used by openshift but project is a openshift term AFAIK

File: cmd/project.go [hunk 3]
```diff
@@ -105,6 +108,38 @@ var projectCreateCmd = &cobra.Command{
 	},
 }
 
+var projectDeleteCmd = &cobra.Command{
+	Use:   "delete",
+	Short: "Delete a project",
+	Long:  "Delete a project and all resources deployed in the project being deleted",
+	Example: `  # Create a new project
```
Comment (syamgk) [inline cmd/project.go]: delete

Comment (anmolbabu) [reply cmd/project.go]: aah my bad :(

**Comment by anmolbabu** [comment]
@kadel @cdrage @surajnarwade @syamgk @mik-dass @cmoulliard Please review..
All comments incorporated and verified

File: Makefile [hunk 4]
```diff
@@ -73,7 +73,7 @@ prepare-release: cross
 
 .PHONY: test
 test:
-	go test -race $(PKGS)
+	go test -gcflags="-N -l" -race $(PKGS)
```
Comment (surajnarwade) [inline Makefile]: can you please add comment about this new flag ? what's the purpose of it ?

File: glide.yaml [hunk 5]
```diff
@@ -36,3 +36,4 @@ import:
   version: v1.4.7
 - package: github.com/olekukonko/tablewriter
   version: d4647c9c7a84d847478d890b816b7d8b62b0b279
+- package: github.com/bouk/monkey
```
Comment (surajnarwade) [inline glide.yaml]: Please mention version for the package, it makes easy to manage dependencies

File: pkg/testingutil/projects.go [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+			getFakeProject("testing"),
+			getFakeProject("prj1"),
+			getFakeProject("prj2"),
+		},
+	}
+}
+
+/*
+func FakeOdoConfig() clientcmd_api.Config {
+	return clientcmd_api.Config{
```
Comment (surajnarwade) [inline pkg/testingutil/projects.go]: Is it WIP, if not we can remove this

File: cmd/project.go [hunk 7]
```diff
@@ -105,6 +108,38 @@ var projectCreateCmd = &cobra.Command{
...
+var projectDeleteCmd = &cobra.Command{
+	Use:   "delete",
+	Short: "Delete a project",
+	Long:  "Delete a project and all resources deployed in the project being deleted",
+	Example: `  # Delete a project
+  odo project delete myproject
+	`,
+	Args: cobra.ExactArgs(1),
+	Run: func(cmd *cobra.Command, args []string) {
+		projectName := args[0]
```
Comment (surajnarwade) [inline cmd/project.go]: we can check here whether given project name exist or not using `project.Exists` ?

Comment (surajnarwade) [reply cmd/project.go]: if don't exist, don't proceed :smile:

File: Makefile (hunk 4 above)
Comment (anmolbabu) [reply Makefile]: Ack

File: pkg/project/project.go [hunk 8]
```diff
@@ -34,6 +38,75 @@ func Create(client *occlient.Client, projectName string) error {
...
+	validProject := false
+	currentProject := GetCurrent(client)
+
+	projects, err := List(client)
+	if err != nil {
+		return errors.Wrapf(err, "unable to fetch list of projects")
+	}
+
+	//Validate the delete requested project
```
[the remark this answers is not in the document]
Comment (anmolbabu) [reply pkg/project/project.go]: ok will use it

File: cmd/project.go (hunk 7 above)
Comment (surajnarwade) [reply cmd/project.go]: it's better to check at early stage ?

Comment (anmolbabu) [reply cmd/project.go]: Not sure how validation in pkg/project is not early stage
I would say any call or invocation out of odo for ex: api calls, cmd invocation may be using oc(I agree we have moved away from this but considering oc just for the purpose of an example) would be the border to consider anything as early or delayed
But I might be missing something. So, fixed it as per the comment..

**Comment by anmolbabu** [comment]
@kadel @cdrage @surajnarwade @syamgk @mik-dass @cmoulliard Please review..
All comments incorporated and verified

File: Makefile [hunk 9]
```diff
@@ -73,7 +73,9 @@ prepare-release: cross
 
 .PHONY: test
 test:
-	go test -race $(PKGS)
+	# Monkey(patching functions for unit tests) sometimes fails to patch a function if inlining is enabled.
+	# Hence disabling inlining with -gcflags="-N -l"
+	go test -gcflags="-N -l" -race $(PKGS)
```
Comment (cdrage) [inline Makefile]: Instead of disabling inlining, why don't we fix it in the code instead? Modifying tests like these may make other issues arise in the future.

File: cmd/project.go (hunk 7 above)
Comment (cdrage) [reply cmd/project.go]: @surajnarwade is right, similar to other checks such as in `create.go` here: https://github.com/redhat-developer/odo/blob/master/cmd/create.go#L92 we should have the Exists code appear in `cmd`.

File: cmd/project.go [hunk 10]
```diff
@@ -105,6 +108,45 @@ var projectCreateCmd = &cobra.Command{
...
+		client := getOcClient()
+
+		// Validate existence of project to be deleted
```
Comment (cdrage) [inline cmd/project.go]: `of the project`

File: cmd/project.go (hunk 7 above)
Comment (cdrage) [reply cmd/project.go]: Nevermind, I see you already added this in line 124

File: cmd/project.go [hunk 11]
```diff
@@ -105,6 +108,45 @@ var projectCreateCmd = &cobra.Command{
...
+		isValidProject, err := project.Exists(client, projectName)
+		checkError(err, "Failed to delete project %s", projectName)
+		if !isValidProject {
+			fmt.Printf("The project %s does not exist. Please check the list of projects using `odo project list`", projectName)
+			os.Exit(1)
+		}
+		var confirmDeletion string
```
Comment (cdrage) [inline cmd/project.go]: add newline above here

File: pkg/project/project.go [hunk 12]
```diff
@@ -34,6 +36,60 @@ func Create(client *occlient.Client, projectName string) error {
...
+
+	//Iterate the project list and see the expected change post deletion
+	for ind, prj := range projects {
+		if prj.Name == projectName {
+			projects = append(projects[:ind], projects[ind+1:]...)
+		}
+	}
+
+	// If there will be any projects post the current deletion,
+	// Randomly choose a project from remainder of project list to set as current
```
Comment (cdrage) [inline pkg/project/project.go]: This is bad, we should pick the next available project in the list, not a random one.

File: pkg/project/project.go [hunk 13]
```diff
@@ -34,6 +36,60 @@ func Create(client *occlient.Client, projectName string) error {
...
+	if len(projects) > 0 {
+		currentProject = projects[0].Name
+	} else {
+		// Set the current project to NOPROJECT
+		currentProject = "NOPROJECT"
```
Comment (cdrage) [inline pkg/project/project.go]: Shouldn't this be blank / just "" like how it currently is implemented with project?

File: pkg/project/project.go [hunk 14]
```diff
@@ -34,6 +36,60 @@ func Create(client *occlient.Client, projectName string) error {
...
+	}
+
+	// Delete the requested project
+	err = client.DeleteProject(projectName)
+	if err != nil {
+		return errors.Wrap(err, "unable to delete project")
+	}
+
+	// If current project is not NOPROJECT, set currentProject as current project
+	if currentProject != "NOPROJECT" {
```
Comment (cdrage) [inline pkg/project/project.go]: blank, not `NOPROJECT`..

File: pkg/project/project_test.go [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+
+			fakeClientSet.ProjClientset.PrependReactor("list", "projects", func(action ktesting.Action) (bool, runtime.Object, error) {
+				return true, testingutil.FakeProjects(), nil
+			})
+
+			fakeClientSet.ProjClientset.PrependReactor("delete", "projects", func(action ktesting.Action) (bool, runtime.Object, error) {
+				return true, nil, nil
+			})
+
+			var guard *monkey.PatchGuard
```
Comment (cdrage) [inline pkg/project/project_test.go]: This is new to me as well as our testing suite. Please comment on what's going on here with "monkey"

File: tests/e2e/cmp_test.go [hunk 16]
```diff
@@ -364,7 +364,11 @@ var _ = Describe("odoCmpE2e", func() {
 	Context("cleaning up", func() {
 		It("should delete the application", func() {
 			runCmd("odo app delete " + appTestName + " -f")
-			runCmd("oc delete project " + projName)
+
+			runCmd("odo project delete " + projName)
+			time.Sleep(15 * time.Second)
```
Comment (cdrage) [inline tests/e2e/cmp_test.go]: adding a sleep timer is bad / adds unnecessary wait. You may have to implement / use some form of ping for this to work. Besides, the project should already be deleted immediately instead of having to wait...

Comment (cdrage) [reply tests/e2e/cmp_test.go]: Our only sleep timer within the e2e tests is here: https://github.com/redhat-developer/odo/blob/c5ffed5dc67b2fcf22a0a7d3a360d34d345bebba/tests/e2e/e2e.go#L27 which is part of ping.

File: tests/e2e/e2e_test.go [hunk 17]
```diff
@@ -471,8 +471,11 @@ var _ = Describe("odoe2e", func() {
 			cmpList := runCmd("odo list")
 			Expect(cmpList).NotTo(ContainSubstring("nodejs"))
 
-			// TODO: `odo project delete` once implemented
-			runCmd("oc delete project " + projName)
+			runCmd("odo project delete " + projName)
+			// Wait 15 secs to see the changes in list of projects post deletion of a project
+			time.Sleep(15 * time.Second)
```
Comment (cdrage) [inline tests/e2e/e2e_test.go]: See other comment regarding adding a timer... This adds unnecessary wait.

File: Makefile (hunk 9 above)
Comment (anmolbabu) [reply Makefile]: Ack...
but inlining can be done by the go compiler this disables inlining by the compiler
But yes, the present code I see currently works even without these flags. So I'll disable them for now we can probably add them when we need it

File: cmd/project.go (hunk 10 above)
Comment (anmolbabu) [reply cmd/project.go]: Ack

File: cmd/project.go (hunk 11 above)
Comment (anmolbabu) [reply cmd/project.go]: Ack

File: pkg/project/project.go (hunk 12 above)
Comment (anmolbabu) [reply pkg/project/project.go]: Hmm we are actually using the first project in list I'll rephrase the words to indicate the same

File: pkg/project/project.go (hunk 13 above)
Comment (anmolbabu) [reply pkg/project/project.go]: Ack

File: pkg/project/project.go (hunk 14 above)
Comment (anmolbabu) [reply pkg/project/project.go]: Ack

File: pkg/project/project_test.go (hunk 15 above)
Comment (anmolbabu) [reply pkg/project/project_test.go]: Ack

File: tests/e2e/e2e_test.go (hunk 17 above)
Comment (anmolbabu) [reply tests/e2e/e2e_test.go]: Ack Done

File: tests/e2e/cmp_test.go (hunk 16 above)
Comment (anmolbabu) [reply tests/e2e/cmp_test.go]: Ack Done

**Comment by cdrage** [comment]
Awesome! Thanks @anmolbabu for this.

I've created an issue to remind ourselves to create a unit test document for this: https://github.com/redhat-developer/odo/issues/717
